### ISSUE TITLE
Tips for how to build QGIS faster as Travis CI under Mac OS?

### ISSUE THREAD

**Issue by michaelkirk** [OSGeo/homebrew-osgeo4mac#78]
Do you have any advice for tuning this recipe to build qgis/master from as quickly as possible?

This might not be the appropriate place for this issue, feel free to redirect me.

Motivation
--------------
I'm trying to get the [OSX build to succeed on travis-ci](https://travis-ci.org/michaelkirk/QGIS), in hopes of getting the new build configuration merged into qgis/master and the osx build status reflected in new pull requests to the project. I'm using this recipe as a guide for configuring the travis node.

The closest I've gotten is probably here: https://travis-ci.org/michaelkirk/QGIS/jobs/49247139

The current problem is that the osx build exceeds travis's 50 minute build time limit (linux build and test cycle on travis is ~25 minutes).

Roughly speaking this recipe takes 15-20 minutes to install all the dependencies (without-globe/without-grass), and then the build takes... well, I don't know because it never finishes, because travis kills it at 50 minutes.

If we assume that compiling on osx has as much output as compiling on linux, a dubious assumption to be sure, then we're nearly done compiling at 50 minutes. We'll need an additional 5ish minutes to run the test suite. And of course, having the mac build take twice as long as the linux build will be a harder sell to upstream.

Any advice?

**Comment by dakcarto** [comment]
First, the many tests failing on Mac need to be updated so they **pass**; here is a [recent test result from my nightly build](http://dash.orfeo-toolbox.org/viewTest.php?onlyfailed&buildid=170649) (though that is based off of Kyngchaos.com supporting libs/frameworks). Then, CI for Mac will make sense to do. I will be updating the PAL labeling tests, which I initially wrote, as they are the most problematic right now on all platforms; however, the other failing tests need work as well.

As an alternative to to using Travis, I have been in contact with the OSGeo about procuring a Mac build server that will run VMware-based VMs of Mac OS X 10.6, .7, .8, .9, .10, for building nightlies, provisioning and running Jenkins, and doing fully bundled Mac releases. Using Travis would be better for CI testing, though, as it would offload those extra CPU cycles.

I already have all of these VMs set up locally on a personal older Mac Pro with 8 cores, which has been retrofitted with cool new SSDs, so QGIS builds much quicker. The OSGeo or QGIS project hardware might be an 8-core Mac Mini i7 (from about a year ago), with two SSD drives in it.

As far as making the build faster, all dependencies should be pre-built and stored on remote location, then pulled in when provisioning the CI VM, instead of being built during a CI build. Only QGIS source should be built, which should not take more than 20 mins, unless it is delegated only 1 CPU (Travis builds get 1.5 CPUs, last I checked). See Travis' [Speeding up the build / Caching the dependencies](http://docs.travis-ci.com/user/speeding-up-the-build/#Caching-the-dependencies).

Can also try building QGIS in CMake `Release` mode, instead of with text stream debug output, though not sure that is the best plan for tracking down build bugs. There are also some unit tests that take waaaay too long on Mac compared to other platforms. So, unit tests need tweaked for whatever is causing that as well.

I have an S3 bucket at http://qgis.dakotacarto.com (where the nightlies are hosted) that can be used for storing dependencies that are auto-built from Homebrew instances on my nightly build box. Such dependency build bots would not always `brew update && brew upgrade` (generically speaking) for updating a CI build's dependency stack, since that might often fail or cause issues with CI stability. CI should always be testing QGIS source builds against a solid dependency stack. This means the Homebrew deps stack needs to be updated _only_ when necessary, and only after testing via a different means.

Creating nightlies based off of Homebrew-only deps is what I plan on doing next. However, I intend those nightlies to be built off of a constantly upgrade Homebrew base, which will help detect incompatibilities with deps changes.

Please, let's revisit this once all Mac unit tests are all passing; otherwise, the CI will always fail.

**Comment by michaelkirk** [comment]
I actually had a slightly different proposal other than fixing all the
tests on mac. In projects with failing test suites, it can be a herculean
effort for essentially one person to fix all the tests. I did start to go
down that road, but figuring out how to get the render tests to pass on
macos in a meaningful way is beyond my experience with the project.

In the past what I've done with large failing test suites is simply and
blindly ignore every test that is currently failing, but from *that point
on* adopt a zero tolerance policy for new failing tests. This is really
only enforceable if you have CI reporting failures on pull requests.

My proposal for achieving that in this code base was using labels thus:
https://github.com/michaelkirk/QGIS/commit/8726df4cc61058a3d2e70201af23fb90733f4ed8

I know this isn't as palatable as "Hey, I fixed your entire test suite for
you!" but unfortunately I am not that good. I do think this is a pragmatic
approach, and it has worked for me in the past. What do you think?

**Comment by dakcarto** [comment]
> In the past what I've done with large failing test suites is simply and blindly ignore every test that is currently failing, but from *that point on* adopt a zero tolerance policy for new failing tests. This is really only enforceable if you have CI reporting failures on pull requests.

I do not think that would be an acceptable approach to many core devs in the project. There was quite an effort to get the majority of Linux unit tests to pass just to have the current CI setup. Until the existing failing tests on Mac are down to a handful (like Linux), it really does not make sense, especially considering the ignored 25+ failing tests would not all be trivial. Indeed, the current test coverage is [quite miserable](http://nyalldawson.net/qgis/coverage/).

First things first: the Mac unit tests need fixed. A false-positive build in CI is just not helpful to anyone; and, QGIS already partially has that with the Linux CI.

**Comment by michaelkirk** [comment]
Thanks for your input @dakcarto! I'll see what I can do.

**Comment by pat-s** [comment]
@dakcarto I would like to integrate a nightly-mac CI to my travis test suite ([example build](https://travis-ci.org/jannes-m/RQGIS/builds/301871864)). 
The easiest way would be to have a bottled nightly that has the same paths as `qgis2` and `qgis-dev` as we are searching within `/usr/local/Cellar` for homebrew installations. 
However, an automated bottling is not that easy in homebrew right? :/

I assume the `.dmg` from http://qgis.dakotacarto.com/ does have other paths? So it would not work if I would download the `.dmg` and install it on travis?

**Comment by dakcarto** [comment]
@pat-s wrote:
> I would like to integrate a nightly-mac CI to my travis test suite (example build).
> The easiest way would be to have a bottled nightly that has the same paths as qgis2 and qgis-dev as we are searching within /usr/local/Cellar for homebrew installations.

We are wanting to do something similar with [the `qgis3-dev` formula](https://github.com/qgis/homebrew-qgisdev/blob/master/Formula/qgis3-dev.rb), where a nightly bottle is built, while at the same time keeping the stable bottle available.

Apparently, you can have a `bottle do ... done` block _inside_ a `devel do ... done` block, so it looks possible, though I have not tested it yet. In this scenario, one can have the `devel` block bottles would be nightlies. The `head do ... done` block should never have bottles.

> However, an automated bottling is not that easy in homebrew right? :/

Yes and no. Used to be easy, but now a bit tricker since Homebrew moved to Jenkins instead of Travis for building bottles. [The `brew-test-bot` script](https://github.com/Homebrew/homebrew-test-bot/tree/master/cmd) is designed to do these types of things. Some routines can be gleaned from it to do so stuff.

I wrote custom bottling routines for this Tap's Travis setup, and need to add auto formula updating to the bottling process.

> I assume the .dmg from http://qgis.dakotacarto.com/ does have other paths? So it would not work if I would download the .dmg and install it on travis?

That's a completely different setup, built off of old Kyngchaos framework installs, and should be avoided. Nothing uses Homebrew.

**Comment by pat-s** [comment]
Thanks @dakcarto for the heads-up!

> Yes and no. Used to be easy, but now a bit tricker since Homebrew moved to Jenkins instead of Travis for building bottles.

Is that the reason why the `qgis3-dev` formula is hosted in `qgis` and not `osgeo4mac` atm?

> Apparently, you can have a bottle do ... done block inside a devel do ... done block, so it looks possible, though I have not tested it yet. In this scenario, one can have the devel block bottles would be nightlies. The head do ... done block should never have bottles.

I would be interested in trying this. Would I have the rights to push to the `osgeo4mac` S3 bucket from my forked repo? Your `deploy` section in `.travis.yml` seems to be public?

By doing a daily CRON travis run with an automated bottling of `qgis3`, the formula could remain as `qgis-dev` and always build a daily bottle of the `qgis` nightly?

**Comment by dakcarto** [comment]
@pat-s wrote:
>> Yes and no. Used to be easy, but now a bit tricker since Homebrew moved to Jenkins instead of Travis for building bottles.
> Is that the reason why the qgis3-dev formula is hosted in qgis and not osgeo4mac atm?

No... well, kinda. 😉 We (QGIS project) were able to get longer build times from folks at Travis for that repo, being part of the QGIS account. The `qgis3-dev` formula is intended to be purely for developers, while any formulae in OSGeo4Mac repo will be released versions.

However, none of that has anything to do with building bottles, per se, except the longer build times. We are considering OSGeo-hosted, but QGIS project-procured, hardware for some of these macOS tasks, i.e. no Travis constraints, with all CI conducted via Jenkins. Travis can be unstable and time-consuming with QGIS builds for macOS.

>> Apparently, you can have a bottle do ... done block inside a devel do ... done block, so it looks possible, though I have not tested it yet. In this scenario, one can have the devel block bottles would be nightlies. The head do ... done block should never have bottles.
> I would be interested in trying this. Would I have the rights to push to the osgeo4mac S3 bucket from my forked repo? Your deploy section in .travis.yml seems to be public?

The Travis setup is public, but the deployment can not be done from a different repo, only the one whose public key has signed the encrypted S3 deployment keys, i.e. you can't push to my S3 bucket from your forked repo.

Also, it is the policy of this repo to not accept bottles with third-party provenance, i.e. PRs will not be accepted that contain third-party bottle source locations.

> By doing a daily CRON travis run with an automated bottling of qgis3, the formula could remain as qgis-dev and always build a daily bottle of the qgis nightly?

Yes, that is the intent, at least for `qgis3-dev`. A standard install will always install the 'nightly' bottle, and `--HEAD` will install tip of dev branch (generally always `master`). Note, once there is a `qgis3` in OSGeo4Mac tap, its `--HEAD` will _not_ do the same, since its branch will be `release-3_x` and not `master`.

**Comment by pat-s** [comment]
> However, none of that has anything to do with building bottles, per se, except the longer build times. We are considering OSGeo-hosted, but QGIS project-procured, hardware for some of these macOS tasks, i.e. no Travis constraints, with all CI conducted via Jenkins. Travis can be unstable and time-consuming with QGIS builds for macOS.

Given this, is there any value in getting the travis build in https://github.com/qgis/homebrew-qgisdev working? If you want to build the bottles on your own hardware and by using Jenkins in the future, the travis build for `qgis3dev` does not seem that important? I did some modifications recently in my [fork](https://travis-ci.org/pat-s/homebrew-qgisdev/builds) (don't look at the branch name :smile: ), e.g. getting rid of the numpy error in your last build. 

It appears to me that you also have increased build times for `osgeo4mac` (qgis builds exceeding one hour). Now I'm confused why you explicitly need the increased build times from the QGIS repo :thinking: :smile: 

BTW: I know about the time limit problems on travis, we have similar problems in the [mlr](https://github.com/mlr-org/mlr) project. 

> The Travis setup is public, but the deployment can not be done from a different repo, only the one whose public key has signed the encrypted S3 deployment keys, i.e. you can't push to my S3 bucket from your forked repo.

Is there any way that I could get involved in the development? I would really like to help and I like solving development issues in my spare time. I do not have a huge experience with `brew` (but it increases) and I also maintain our RQGIS [travis build](https://github.com/jannes-m/RQGIS/blob/master/.travis.yml).
